### ISSUE TITLE
Problem with running in sequence - Running version() after other tasks finished

### ISSUE THREAD

**Issue by mgussekloo** [laravel/elixir#126]
I'm having trouble with running Elixir's version() after all the other tasks are completed. I get errors because my tasks haven't finished yet and version() is already being called. Elixir should be able to handle the use-case, so I'm probably doing something wrong. Help!

Gulp-file:
```javascript
    // promo
    mix.sass("promo.scss")
    .scripts([
        "js/promo.js",
    ], "public/js/promo.js", "resources/assets/")

    // main site
    .sass("main.scss", "resources/assets/css/")
    .styles([
    	"css/main.css",
    ], "public/css/main.css", "resources/assets/")
    .scripts([
        "js/main.js",
        "glide/glide.js",
        "picturefill/picturefill.js",
    ], "public/js/main.js", "resources/assets/")
    .copy("resources/assets/js/jquery-1.7.1.min.js", "public/js/jquery-1.7.1.min.js")
	.version([
		"css/promo.css", "js/promo.js",
    	"css/main.css", "js/main.js"
    ]);
```

Result:
```
[17:32:56] Starting 'sass'...
[17:32:56] Running Sass: resources/assets/sass/promo.scss
[17:32:56] Running Sass: resources/assets/sass/main.scss
[17:32:57] Finished 'sass' after 692 ms
[17:32:57] Starting 'version'...
[17:32:57] Versioning: public/css/promo.css,public/js/promo.js,public/css/main.css,public/js/main.js
[17:32:57] Starting 'styles'...
[17:32:57] Merging: resources/assets/css/main.css

events.js:72
        throw er; // Unhandled 'error' event
              ^
Error: ENOENT, lstat '/home/vagrant/Sites/Insight/public/css/main.css'
```

As you can see, Sass() is called, then Version() is called, and finally Styles() is called.
This is wrong because correct output of Version() depends on Styles() being completed.

Any help is appreciated.

**Comment by JeffreyWay** [comment]
Hmm - are you using the latest version of Elixir? Have you done `npm update` lately?

**Comment by mgussekloo** [comment]
Hi Jeffrey, 

Thanks for your reply! (And for all your work on this project.) 

Yes, I saw some of the issues on here about concurrency so I updated using npm update. The Elixir package.json now says "name": "laravel-elixir", "version": "0.18.7".

Some more info: removing the ```//promo``` part from the gulpfile apparently resolves problem. So I've updated my gulpfile to:

```javascript
    // main site
   mix.sass("main.scss", "resources/assets/css/")
    .styles([
        "css/main.css",
    ], "public/css/main.css", "resources/assets/")
    .scripts([
        "js/main.js",
        "glide/glide.js",
        "picturefill/picturefill.js",
    ], "public/js/main.js", "resources/assets/")
    .copy("resources/assets/js/jquery-1.7.1.min.js", "public/js/jquery-1.7.1.min.js")
    .version([
        "css/promo.css", "js/promo.js",
        "css/main.css", "js/main.js"
    ]);
```

And the output is a lot more stable: (the js rules are missing because I'm testing this by saving a .scss file).

```
[06:53:37] Starting 'sass'...
[06:53:37] Running Sass: resources/assets/sass/main.scss
[06:53:37] Finished 'sass' after 865 ms
[06:53:38] Starting 'styles'...
[06:53:38] Merging: resources/assets/css/main.css,resources/assets/glide/css/glide.css
[06:53:38] Finished 'styles' after 63 ms
[06:53:38] Starting 'version'...
[06:53:38] Versioning: public/css/promo.css,public/js/promo.js,public/css/main.css,public/js/main.js
[06:53:38] Finished 'version' after 211 ms
```

Let me know if there is anything I can do to debug this issue!

**Comment by tjort** [comment]
I'm wrestling with the same issue as well. Very interested in a potential fix.

**Comment by mgussekloo** [comment]
I'm still getting the error. For some reason it's firing version() before finishing scripts(), and then running version() again afterwards?

```javascript

    mix

    .copy("bower_components/jquery/jquery.js", "resources/assets/vendor/js/jquery.js")

    .copy("bower_components/flickity/dist/flickity.pkgd.js", "resources/assets/vendor/js/flickity.js")
    .copy("bower_components/flickity/dist/flickity.css", "resources/assets/vendor/css/flickity.css")

    .copy("bower_components/picturefill/dist/picturefill.js", "resources/assets/vendor/js/picturefill.js")

    .copy("bower_components/super-gradient/_super-gradient.scss", "resources/assets/vendor/sass/_super-gradient.scss")

    .sass("main.scss", "resources/assets/css", "resources/assets/sass")

    .styles([
    	"vendor/css/flickity.css",
    	"css/main.css"
	], "public/css/main-all.css", "resources/assets")

	.scripts([
		"vendor/js/jquery.js",
		"vendor/js/picturefill.js",
		"vendor/js/flickity.js",
		"js/main.js"
	], "public/js/main-all.js", "resources/assets")

	.version([
    	"css/main-all.css", "js/main-all.js"
    ]);
```

```
[12:51:40] Starting 'scripts'...
[12:51:40] Merging: resources/assets/vendor/js/jquery.js,resources/assets/vendor/js/picturefill.js,resources/assets/vendor/js/flickity.js,resources/assets/js/main.js
[12:51:40] Starting 'version'...
[12:51:40] Versioning: public/css/main-all.css,public/js/main-all.js

-> [12:51:40] File not found: public/js/main-all.js <-

[12:51:40] Finished 'scripts' after 213 ms
[12:51:40] Finished 'version' after 44 ms
[12:51:41] Starting 'version'...
[12:51:41] Versioning: public/css/main-all.css,public/js/main-all.js
[12:51:41] Finished 'version' after 66 ms
```

**Comment by shehi** [comment]
Same problem with me as well. In my case, here:
https://github.com/AudithSoftworks/Basis/blob/admin-panel-ui/gulpfile.js
I have created a new extension called ```compass``` which utilizes gulp-compass library. Below, at 
https://github.com/AudithSoftworks/Basis/blob/admin-panel-ui/gulpfile.js#L56
although I have ```compass()``` task at the top, it's being carried out last, messing ```version()```'ing.

**Comment by shehi** [comment]
@JeffreyWay : Is there any development ongoing for this issue? If not, I will have to switch to Gulp instead, and do what official Gulp recipe recommends https://github.com/gulpjs/gulp/blob/master/docs/recipes/running-tasks-in-series.md . Thanks.

**Comment by laracasts** [comment]
I haven't had a chance to really dig into it just yet. But PRs are welcomed. Otherwise, I'll try to review it after the holiday.

**Comment by shehi** [comment]
@laracasts @JeffreyWay : I noticed you already have ```run-sequence``` incorporated into L5 Elixir. But, as reported above, it doesn't somehow work. Does that mean orchestrator has been updated so that the hack (run-sequence) doesn't work anymore?

**Comment by htulibacki** [comment]
I also having this problem.

I observed, that when I'm versioning many files, version task starts before less task is completed. Sadly, I'm currently unable to use gulp watch at all. This problem never appears, when i'm using gulp command.

I'm using versio 2.1.1.

**Comment by spajz** [comment]
Same problem with me as well. I have latest Elixir version.
![elixir](https://cloud.githubusercontent.com/assets/801824/7909498/e1bc913a-084c-11e5-86da-65eacae55464.png)

**Comment by JeffreyWay** [comment]
I'm having trouble reproducing his.

If no one will submit a PR, the next most helpful thing is if I could receive a barebones example that I can  try out on my end, to see the issue firsthand. If you have one, either reply here, or create a new issue please.

**Comment by shehi** [comment]
@JeffreyWay : and why would you close this one anyway?! It's a case reproduced by many people.

As for the case: 

1) add some SCSS file to a project
2) bring in gulp-compass lib with bower
3) extend elixir with new task for gulp-compass and parse scss files with it (clearly you will need compass installed btw)
4) and add versioning after compass task in your gulpfile.js
5) gulp it: you will see that compass task runs AFTER versioning

P.S. you can use gulp-sass if u desire, but as you already know it doesn't support sass-globbing.

**Comment by spajz** [comment]
I can no longer help with debug. I do not use Elixir anymore with Laravel.

**Comment by mgussekloo** [comment]
I'm not sure closing the issue will get it fixed any faster.

I've tested a bit more and it seems the problem is related to the watch command. Maybe run-sequence is called on the default command, but not called on subsequent "watch" commands? Just to be sure I've added some console logging to the dependency "run-sequence" of Elixir and I can see run-sequence activate and watching events the first time (non-watched), but not the second time (watched).

It's not a PR, but it's a start.

**Comment by phaza** [comment]
Seems weird to close an unresolved bug just because you can't reproduce, @JeffreyWay.
This is a problem I too struggle with, maybe it's related to the compass or sass script?
Definitely some async problem. I can see that compass is starting before version, but version is starting before compass is finished (and actually finishes before too).

**Comment by shehi** [comment]
I concur with @phaza

**Comment by bhulsman** [comment]
I'm having the same problem; 'version' runs before 'scripts' finished.

**Comment by AppSynergy** [comment]
An example:    

* see last 5 lines - version starts before styles, so the files it's trying to version temporarily don't exist?
* seems like a race condition, if I just start saving a sass file repeatedly, it'll work N times before hitting ENOENT (*nothing to do with tree-folk playing ambient, it turns out..*)

```bash
$ gulp watch
[15:50:48] Using gulpfile /{PROJ}/gulpfile.js
[15:50:48] Starting 'watch'...
[15:50:48] Starting 'coffee'...
[15:50:48] Finished 'watch' after 80 ms
[15:50:48] gulp-notify: [Laravel Elixir] CoffeeScript Compiled!
[15:50:48] Finished 'coffee' after 199 ms
[15:50:48] Starting 'sass'...
[15:50:48] Running Sass: resources/assets/sass/**/*.+(sass|scss)
[15:50:49] gulp-notify: [Laravel Elixir] Sass Compiled!
[15:50:49] Finished 'sass' after 449 ms
[15:50:49] Starting 'styles'...
[15:50:49] Merging: public/css/backend.css,public/css/vendor.css
[15:50:49] Merging: public/css/frontend.css,public/css/vendor.css
[15:50:49] Finished 'styles' after 66 ms
[15:50:49] Starting 'version'...
[15:50:49] Versioning: public/css/frontend-bs.css,public/css/backend-bs.css
[15:50:49] Finished 'version' after 28 ms
[15:50:49] Starting 'copy'...
[15:50:49] Finished 'copy' after 23 ms
[15:50:49] Starting 'watch-assets'...
[15:50:49] Finished 'watch-assets' after 23 ms
[15:50:53] Starting 'sass'...
[15:50:53] Running Sass: resources/assets/sass/**/*.+(sass|scss)
[15:50:53] Starting 'styles'...
[15:50:53] Merging: public/css/backend.css,public/css/vendor.css
[15:50:53] Merging: public/css/frontend.css,public/css/vendor.css
[15:50:53] gulp-notify: [Laravel Elixir] Sass Compiled!
[15:50:53] Finished 'sass' after 151 ms
[15:50:53] Finished 'styles' after 28 ms
[15:50:53] Starting 'version'...
[15:50:53] Versioning: public/css/frontend-bs.css,public/css/backend-bs.css
[15:50:53] Starting 'styles'...
[15:50:53] Merging: public/css/backend.css,public/css/vendor.css
[15:50:53] Merging: public/css/frontend.css,public/css/vendor.css

events.js:72
        throw er; // Unhandled 'error' event
              ^
Error: ENOENT, lstat '/var/www/equas/public/css/frontend-bs.css'
```

```js
var elixir = require('laravel-elixir');
require('laravel-elixir-bower');

elixir(function(mix) {

    mix.coffee('','',{ bare: true });

    mix.sass();
    
    mix.bower();

    mix.styles(
        [
            'backend.css',
            'vendor.css'
        ],
        'public/css/backend-bs.css',
        'public/css'
    ).styles(
        [
            'frontend.css',
            'vendor.css'
        ],
        'public/css/frontend-bs.css',
        'public/css'
    );

    mix.version(['css/frontend-bs.css', 'css/backend-bs.css']);

    mix.copy('public/fonts', 'public/build/fonts');
});
```

**Comment by AppSynergy** [comment]
It seems I can at least avoid the errors with a naive and hacky implementation of [Gulp task dependancies](https://github.com/gulpjs/gulp/blob/master/docs/recipes/running-tasks-in-series.md), along the lines of:

```js
var buildTask = function(src, buildDir, deps) {
    buildDir = getBuildDir(buildDir);
    deps = typeof deps !== 'undefined' ?  deps : [];
    
    gulp.task('version', deps, function() {
...
```
#### Reasons why this probably isn't the answer

- If I make `version` depend on `styles`, I seem to end up with `styles` executing 3 times, and `version` twice (each one more time than I'd expected!) - but at least they do run in series (seems to reliably go `s-s-v-s-v` - puzzling...)
- I'm working with Elixir v1.x, due to #160

**Comment by frankmichel** [comment]
I am not sure why this issue is closed. I am having the exact same problem. `gulp watch` constantly crashes while a regular `gulp` works without any problems. If I comment out the versioning, gulp watch runs without any issues.

**Comment by bhulsman** [comment]
@frankmichel I managed to reduce the number of occurrences by not chaining it al together. So instead of...

```
mix
.script()
.less()
.version();
```

I did something like...

```
mix.script();
mix.less();
mix.version();
```

It's just an example, not working at the project at the moment, but it should give you an idea on how to work around the issue. In my case it worked quite well anyway.

**Comment by frankmichel** [comment]
@bhulsman Thanks for your quick reply! Unfortunately this is already how I have set it up. And yet the errors occur very frequently. My set-up looks like this:

```
elixir(function(mix) {

    mix.sass('app.backend.scss', 'public/css/app.backend.css');
    mix.sass('app.frontend.scss', 'public/css/app.frontend.css');

    mix.scripts([

        'plugins/some.js',
        'plugins/more.js',
        'plugins/files.js',

    ], 'public/js/app.min.js');

    mix.copy('resources/lang/app.language.*.js', 'public/js');

    mix.browserify('app.main.js', 'public/js/app.main.js');

    mix.version(['public/css', 'public/js']);
});
```

**Comment by Gouvi** [comment]
I too am receiving the same issue:

``` javacript
var gulp        = require('gulp');
var elixir      = require('laravel-elixir');
var livereload  = require('laravel-elixir-livereload');

/*
 |--------------------------------------------------------------------------
 | Elixir Asset Management
 |--------------------------------------------------------------------------
 |
 | Elixir provides a clean, fluent API for defining some basic Gulp tasks
 | for your Laravel application. By default, we are compiling the Sass
 | file for our application, as well as publishing vendor resources.
 |
 */

elixir(function(mix) {
    mix.sass('style.scss');

    mix.scripts('scripts.js');

    mix.version(['public/css/style.css', 'public/js/scripts.js']);

    mix.livereload();
});
```

**Comment by apancutt** [comment]
It's a race condition.

Sequence of events:

1. Modify `resources/assets/js/app.js`
2. The `scripts` task will be invoked by the watcher on `resources/assets/js/app.js`
3. The `scripts` task creates a new file at the destination path (`public/js/all.js`)
4. The `version` task will be invoked by the watcher on `public/js/all.js`
5. The `version` task reads `public/js/all.js` to obtain the MD5 sum (the version hash)
6. The `version` task creates the necessary files in the build directory (`public/build`)

That all works fine until a subsequent change is made to `resources/assets/js` in quick succession.

If step 5 has not completed on the first iteration and step 3 is in progress on the secondary iteration (i.e. `public/js/all.js` has been deleted but not yet re-created) you'll get the errors described above.

It's hard to give guaranteed reproduction steps as it ultimately comes down the rate of making changes to asset files and the speed your CPU churns through the tasks, but hopefully I've given enough detail to assist further investigation.

Unfortunately, I'm not familiar enough with gulp, gulp-batch or elixir to submit a PR myself.

**Comment by chintankotadia** [comment]
I had a similar issue but got resolved from solution posted on Stackoverflow (https://stackoverflow.com/questions/16748737/grunt-watch-error-waiting-fatal-error-watch-enospc).

Hope, this helps someone! ;)

**Comment by alexw23** [comment]
I fixed the issue, don't reference scripts as a folder in `version()`, instead reference it as a path to the script.

Do this:

```
.version([
            "css/app.css",
            "scripts/script.js",
            "scripts/modules.js",
            "fonts",
            "icons",
            "images"
        ]);
```

Instead of


```
.version([
            "css",
            "scripts",
            "fonts",
            "icons",
            "images"
        ]);
```

**Comment by briedis** [comment]
@alexw23 That's cool, but we usually have more than 2 files to version...

**Comment by frankmichel** [comment]
@alexw23 I never used explicit script references, always paths:

`mix.version(['public/css', 'public/js']);`

And yet the race condition occurred...

**Comment by alexw23** [comment]
@frankmichel that's what I'm saying - be explicit and it won't happen (don't use paths).

**Comment by frankmichel** [comment]
@alexw23 aaaaaah sorry it's early in the morning here :) i'll give that a try! :+1:

**Comment by claudiodekker** [comment]
I've spent some time trying to figure out what's wrong, and I think I've found the culprit.
In `laravel-elixir/dist/tasks/watch.js`, there's this piece of code:
```javascript
_gulp2.default.watch(task.watchers, { interval: 1000 }, batch(batchOptions, function (events) {
    events.on('end', _gulp2.default.start(task.name));
}));
```

What happens here, is that `gulp-watch` polls the watched file every second, and if something's changed, it triggers the third param, which is a callback method. [You can read the documentation for this here](https://www.npmjs.com/package/gulp-watch#watchglob-options-callback).

Now, this callback method, which is a call to the `batch` node module [(Github link)](https://github.com/floatdrop/gulp-batch#batchoptions-callback-errorhandler), is that is has a timeout value with a default of 100ms. 

Fortunately for most of the normal users, @JeffreyWay updated this timeout to 1000ms (equal to `watch` polling time), assuming an CSS asset wouldn't take more than a second to compile. You can see this in `laravel-elixir/dist/Config.js`:
````javascript
    /*
     |----------------------------------------------------------------
     | File System Event Batching
     |----------------------------------------------------------------
     |
     | You likely won't need to modify this object. That said, should
     | you need to, these settings are exclusive to the watch task.
     | They set the limit and timeout for running batch-updates.
     |
     */

    batchOptions: {
        // https://github.com/floatdrop/gulp-batch#batchoptions-callback-errorhandler
        limit: undefined,
        timeout: 1000
    },
```

Locally, I've updated this value to 20000 (20 seconds), which successfully averted this issue. However, the downside is that tasks are now waiting ages (20 seconds ¯\_(ツ)_/¯) before continuing on to the other tasks (version, browserSync..)

Once again, I'm not sure if this is the final part of the puzzle, but there's a chance it might be. I'll look into this some more tomorrow morning, but if anyone's interested to give this a shot and figure out some more, then please do. The sooner this is resolved, the better.

**Comment by claudiodekker** [comment]
Alright, so, did some research and here's the **definite cause** of this bug.

## The Logic

Apparently, `version` adds all passed files to the 'watch' list, like all other tasks do (`scripts`, `styles`, `sass`, ...). This means that if one of the watched ('to be versionized') files gets updated, it will trigger the version task.

That said, there's a fair chance you'll say "_that's nonsense!_", because this would mean it would trigger _all the time_, with _pretty much every file_. A partial truth to this is that it actually **does** trigger with every file.

The other part of that same truth, is that Elixir's watch task uses [this node module ('batch')](https://github.com/floatdrop/gulp-batch#batchoptions-callback-errorhandler) to prevent the triggering of such tasks in rapid succession. Here's how it's being used in `laravel-elixir/dist/tasks/watch.js`/;

```javascript
var batchOptions = _laravelElixir2.default.config.batchOptions;

if (task.hasWatchers()) {
     _gulp2.default.watch(task.watchers, { interval: 1000 }, batch(batchOptions, function (events) {
        events.on('end', _gulp2.default.start(task.name));
    }));
}
```

What the snippet above does, is poll each watched file (`task.watchers`) every `1000`ms, and if there's been a change in the file, it would hit the callback (`batch(batchOptions, function (events) {...`)

As you can see, batchOptions is being passed, which is the following Config array (`laravel-elixir/dist/Config.js`):
```javascript
    /*
     |----------------------------------------------------------------
     | File System Event Batching
     |----------------------------------------------------------------
     |
     | You likely won't need to modify this object. That said, should
     | you need to, these settings are exclusive to the watch task.
     | They set the limit and timeout for running batch-updates.
     |
     */

    batchOptions: {
        // https://github.com/floatdrop/gulp-batch#batchoptions-callback-errorhandler
        limit: undefined,
        timeout: 1000
    },
```

Basically, instead of doing the (default) `100`ms timeout, @JeffreyWay's changed it to `1000`ms to account for most of `scripts`, `styles` causing this exact behavior.

## The Cause ('Race Condition')

So, now that you understand how Elixir's `watch` task works, you've probably already guessed why this happens. If it doesn't, let me explain why:

When you're compiling many scripts (30-40), or scripts that the compiler has to take a bit longer for, response time will change. Here's an example:

```
[22:41:22] gulp-notify: [Laravel Elixir] Stylesheets Merged!
[22:41:22] gulp-notify: [Laravel Elixir] Stylesheets Merged!
[22:41:22] gulp-notify: [Laravel Elixir] Stylesheets Merged!
[22:41:22] gulp-notify: [Laravel Elixir] Stylesheets Merged!
[22:41:22] gulp-notify: [Laravel Elixir] Stylesheets Merged!
[22:41:22] gulp-notify: [Laravel Elixir] Stylesheets Merged!
[22:41:22] gulp-notify: [Laravel Elixir] Stylesheets Merged!
[22:41:22] gulp-notify: [Laravel Elixir] Stylesheets Merged!
[22:41:22] gulp-notify: [Laravel Elixir] Stylesheets Merged!
[22:41:22] gulp-notify: [Laravel Elixir] Stylesheets Merged!
[22:41:22] gulp-notify: [Laravel Elixir] Stylesheets Merged!
[22:41:22] gulp-notify: [Laravel Elixir] Stylesheets Merged!
[22:41:22] gulp-notify: [Laravel Elixir] Stylesheets Merged!
[22:41:22] gulp-notify: [Laravel Elixir] Stylesheets Merged!
[22:41:22] gulp-notify: [Laravel Elixir] Stylesheets Merged!
[22:41:22] gulp-notify: [Laravel Elixir] Stylesheets Merged!
[22:41:23] gulp-notify: [Laravel Elixir] Stylesheets Merged!
[22:41:23] gulp-notify: [Laravel Elixir] Stylesheets Merged!
[22:41:23] gulp-notify: [Laravel Elixir] Stylesheets Merged!
[22:41:23] gulp-notify: [Laravel Elixir] Stylesheets Merged!
[22:41:26] gulp-notify: [Laravel Elixir] Stylesheets Merged!
[22:41:26] gulp-notify: [Laravel Elixir] Stylesheets Merged!
[22:41:27] gulp-notify: [Laravel Elixir] Stylesheets Merged!
[22:41:27] gulp-notify: [Laravel Elixir] Stylesheets Merged!
[22:41:27] gulp-notify: [Laravel Elixir] Stylesheets Merged!
[22:41:31] gulp-notify: [Laravel Elixir] Stylesheets Merged!
[22:41:35] gulp-notify: [Laravel Elixir] Stylesheets Merged!
[22:41:40] gulp-notify: [Laravel Elixir] Stylesheets Merged!
```

Note the timestamp on the last few lines. As you can see, there's multiple occassions this will trigger the version task. The first time this happens, will be at the very top. Let's take the first two lines as example:
```
[22:41:22] gulp-notify: [Laravel Elixir] Stylesheets Merged!
[22:41:22] gulp-notify: [Laravel Elixir] Stylesheets Merged!
``` 

Because these files have been changed within one second (`1000`ms), the timer on the `batch` task gets reset. This happens for most other files, however, for the following (bold) lines, it will actually go ahead and compile `version` tasks:

```
[22:41:22] gulp-notify: [Laravel Elixir] Stylesheets Merged!
[22:41:22] gulp-notify: [Laravel Elixir] Stylesheets Merged!
**[22:41:23] gulp-notify: [Laravel Elixir] Stylesheets Merged!**
[22:41:26] gulp-notify: [Laravel Elixir] Stylesheets Merged!
[22:41:26] gulp-notify: [Laravel Elixir] Stylesheets Merged!
[22:41:27] gulp-notify: [Laravel Elixir] Stylesheets Merged!
[22:41:27] gulp-notify: [Laravel Elixir] Stylesheets Merged!
**[22:41:27] gulp-notify: [Laravel Elixir] Stylesheets Merged!**
**[22:41:31] gulp-notify: [Laravel Elixir] Stylesheets Merged!**
**[22:41:35] gulp-notify: [Laravel Elixir] Stylesheets Merged!**
**[22:41:40] gulp-notify: [Laravel Elixir] Stylesheets Merged!**
```

As you can imagine, the result of this in your actual console, would look something like this:
![screen shot 2016-05-26 at 01 08 40](https://cloud.githubusercontent.com/assets/1752195/15558943/7509f318-22de-11e6-80bf-1e3170957ddf.png)

**Comment by JeffreyWay** [comment]
Thanks for the good description of the issue. 

Maybe we can just set this up where mix.version() no longer adds a watcher. It just triggers once gulp is finished. Would that be a problem?

**Comment by claudiodekker** [comment]
I'm not sure that could be done (easily), because there isn't really a way to tell gulp is 'done' in a watch process, I think.

Besides, browserSync can still be active.

**Comment by claudiodekker** [comment]
@JeffreyWay I've gotten a working fix, but I can't find a branch to fork from & the EC6 branch seems broken?:

```sh
vagrant@homestead:~/tinker$ gulp
/home/vagrant/tinker/node_modules/laravel-elixir/dist/index.js:40
Elixir.Task = require('./Task')(Elixir);
                               ^

TypeError: require(...) is not a function
```

**Comment by bahjaat** [comment]
Interesting fix there. Hope it will get merged shortly.

**Comment by AppSynergy** [comment]
A tip for anyone still struggling with their tasks:  I've since started using npm scripts instead of gulp/elixir, and not looked back!

**Comment by claudiodekker** [comment]
> Interesting fix there. Hope it will get merged shortly.
Honestly, I hope the new version of Elixir (with webpack, etc.) solves this issue on a structural level. I think @JeffreyWay rebuilt it from the ground up, although I'm not entirely sure.

> A tip for anyone still struggling with their tasks: I've since started using npm scripts instead of gulp/elixir, and not looked back!

That works, I suppose, but the entire thing that Laravel Elixir tries to solve is the ease of use for doing these tasks.. Using something else kind of defeats the purpose. Although it does _work around_ the issue, I wouldn't say it's a great tip for Elixir users ;)

**Comment by briedis** [comment]
@AppSynergy 
I rewrote a vanilla gulpfile with 190 lines to about 50 lines with elixir.

**Comment by AppSynergy** [comment]
@claudiodekker  
Nope, I guess not. 

Maybe the point should be:  npm scripts could work well _with_ Elixir in certain situations.  This issue was a dealbreaker for me last year, and this was the direction I went in instead. Running tasks in parallel/series with traditional (and portable!) notation like:
````
"scripts": {
    "AthenB": "A && B",
    "AandBandC" : "A & B & C",
````

works well for me. 

190 -> 50 lines is great however you do it, of course!

+1 for new version with rewrites and webpack, sounds interesting.

**Comment by joseph-d** [comment]
This issue still seems to be occurring in Elixir ^6.0.0-9

**Comment by Venu-k** [comment]
@JeffreyWay, I too have the same issue, when I run gulp watch it breaks randomly with version(), I think the problem is with manifest file, while running version multiple times, it is checking for the file which is already deleted. please can any one help me to solve this issue.

**Comment by a3diti** [comment]
Hello guys,
**I fixed my issue**
-
My setup is like this:
<img width="611" alt="screen shot 2016-09-29 at 00 00 03" src="https://cloud.githubusercontent.com/assets/3646190/18934007/b9c8534e-85d7-11e6-98cb-1fb9e925f875.png">
&nbsp;
My error was like this:
<img width="985" alt="screen shot 2016-09-28 at 23 53 34" src="https://cloud.githubusercontent.com/assets/3646190/18933815/d84b7ef0-85d6-11e6-857f-f91de553fbef.png">
&nbsp;
The fix worked by doing this:
<img width="466" alt="screen shot 2016-09-28 at 23 55 07" src="https://cloud.githubusercontent.com/assets/3646190/18933875/100bdf4c-85d7-11e6-9e83-92451d4c4395.png">
&nbsp;
It seems you need to delete old files and folders: **rev-manifest.json, js/** and **css/** from **build** folder
&nbsp;
just run
```
cd project/public/build/
rm -rf *
```

I hope it helps since it was a little hard for me as a beginner on laravel, keep up and never give up trying ;)

**Comment by Venu-k** [comment]
`process.env.DISABLE_NOTIFIER = true;
var gulp         = require('gulp');
var gulpFilter   = require('gulp-filter');
var rename       = require('gulp-rename');
var elixir = require('laravel-elixir');
var del = require('del');
var watch = require('gulp-watch');
require('laravel-elixir-js-uglify');
var uglifycss = require('gulp-uglifycss');
elixir.config.sourcemaps = false;
/*
 |--------------------------------------------------------------------------
 | Elixir Asset Management
 |--------------------------------------------------------------------------
 |
 | Elixir provides a clean, fluent API for defining some basic Gulp tasks
 | for your Laravel application. By default, we are compiling the Less
 | file for our application, as well as publishing vendor resources.
 |
 */
elixir.extend('uglifycss', function(src, outputPath, baseDir, options) {


  new elixir.Task('uglifycss', function() {
    return gulp.src('public/css/*.css')
      .pipe(uglifycss({
        "uglyComments": true
      }))
      .pipe(gulp.dest('public/css'));
  })
  // Register watcher for source path
    .watch('resources/assets/**/*.css');

});



gulp.task('minifyCSS', function() {
  elixir(function (mix) {
    /* Minify and merge Scripts */
    mix.styles([
        'template/plugins/jquery-ui/css/jquery-ui.css',
        'template/stylesheet/bootstrap.css',
        'template/stylesheet/layout.css',
        'template/stylesheet/layout.css',
        'template/stylesheet/uielement.css',
        'template/plugins/fileupload/css/fileupload.css',
        'template/plugins/ladda/css/ladda.css',
        'template/plugins/owl/css/owl-carousel.css',
        'template/plugins/layerslider/css/layerslider.css',
        'template/stylesheet/themes/layouts/fixed-sidebar.css',
        'template/plugins/selectize/css/selectize.css',
        'template/plugins/jquery-ui/css/jquery-ui.css',
        'template/plugins/datatables/css/datatables.css',
        'template/plugins/datatables/css/tabletools.css',
        'template/plugins/gritter/css/gritter.css'
      ], 'public/css/template.min.css')
      .styles([
        'jquery.fileupload.css',
        'jquery.fileupload-ui.css'
      ], 'public/css/vendor.min.css')
      .styles([
        'rm-theme.css',
        'rm.css',
        'custom.css'
      ], 'public/css/rm.min.css');
    mix.uglifycss();
    mix.version([
      'css/**',
      'js/**']);
  });
});

gulp.task('minifyJS', function() {
  elixir(function (mix) {
    /* Minify and merge Scripts */
    mix.scripts([
      '/template/javascript/vendor.js',
      'angular/angular.js',
      'angular/angular-route.js',
      '/template/javascript/core.js',
      '/template/javascript/backend/app.js',
      '/template/javascript/frontend/app.js',
      '/template/javascript/frontend/home/home-v1.js',
      'bootstrap/ui-bootstrap-0.12.1.min.js',
      'pusher/pusher.min.js',
      'angular/pusher-angular.min.js'], 'public/js/core.min.js');
    mix.scripts(['/template/plugins/jquery-ui/js/jquery-ui.js',
      '/template/plugins/jquery-ui/js/addon/timepicker/jquery-ui-timepicker.js',
      '/template/plugins/jquery-ui/js/jquery-ui-touch.js',
      '/template/plugins/selectize/js/selectize.js',
      '/template/plugins/datatables/js/jquery.dataTables.js',
      '/template/plugins/datatables/tabletools/js/dataTables.tableTools.js',
      '/template/plugins/datatables/js/datatables-bs3.js',
      '/template/javascript/backend/tables/datatable.js',
      '/template/plugins/ladda/js/vendor/spin.js',
      '/template/plugins/ladda/js/ladda.js',
      '/template/plugins/fileupload/js/vendor/jquery.ui.widget.js',
      '/template/plugins/fileupload/js/vendor/load-image.js',
      '/template/plugins/fileupload/js/vendor/load-image-meta.js',
      '/template/plugins/fileupload/js/vendor/load-image-exif.js',
      '/template/plugins/fileupload/js/vendor/load-image-ios.js',
      '/template/plugins/fileupload/js/vendor/canvas-to-blob.js',
      '/template/plugins/fileupload/js/vendor/jquery.iframe-transport.js',
      '/template/plugins/fileupload/js/jquery.fileupload.js',
      '/template/plugins/fileupload/js/jquery.fileupload-process.js',
      '/template/plugins/fileupload/js/jquery.fileupload-image.js',
      '/template/plugins/fileupload/js/jquery.fileupload-audio.js',
      '/template/plugins/fileupload/js/jquery.fileupload-video.js',
      '/template/plugins/fileupload/js/jquery.fileupload-validate.js',
      '/template/plugins/fileupload/js/jquery.fileupload-ui.js',
      '/template/javascript/backend/forms/fileupload.js',
      '/template/plugins/gritter/js/jquery.gritter.js',
      '/template/javascript/backend/components/notification.js',
      '/template/plugins/owl/js/owl.carousel.js',
      '/template/plugins/modernizr/js/modernizr.js'], 'public/js/template.min.js');
    mix.scripts(['angular/angular-sanitize.min.js',
      'angular/angular-ui-date.js',
      'angular/angular-selectize.js',
      'angular/dimensions.min.js',
      'angular/tooltip.min.js',
      'angular/date-formatter.min.js',
      'angular/date-parser.min.js',
      'angular/datepicker.min.js',
      'angular/timepicker.min.js',
      'angular/datepicker.tpl.js',
      'angular/tooltip.tpl.js',
      'angular/timepicker.tpl.js',
      'angular/angular-jquery.fileupload.js'], 'public/js/vendor.min.js');
    mix.scripts([
        'rm/rmapp.js',
        'controllers/fileUploadController.js',
        'controllers/helperFunctions.js',
        'controllers/site/siteController.js',
        'controllers/site/signUpController.js',
        'controllers/meeting/setupMeetingController.js',
        'controllers/meeting/recordMinutesController.js',
        'controllers/meeting/meetingTeamController.js',
        'controllers/meeting/detailsController.js',
        'controllers/meeting/notesDetailsController.js',
        'controllers/meeting/setRecurrenceController.js',
        'controllers/meeting/editRecurrenceController.js',
        'controllers/meeting/viewInProgress.js',
        'controllers/user/userController.js',
        'controllers/meeting/rmDetailsController.js',
      ], 'public/js/app.min.js')
      .scripts([
        'rm/custom.js'
      ], 'public/js/rm.min.js');
    mix.uglify('**/*.js', 'public/js', 'public/js', {
      mangle: false
    });
    mix.version([
      'css/**',
      'js/**']);
  });
});

/* 'gulp version' - Run after each change in dev */
gulp.task('version', function() {
  elixir(function (mix) {
    mix.version([
      'css/**',
      'js/**']);
  });
});

/* 'gulp clean' - Run when you want to clean the minified and build files */
gulp.task('clean', function() {
  return del(['public/build/**', 'public/js/**', 'public/css/**']);
});

/* 'gulp copy' - Run only when build folder is deleted */
gulp.task('copy', function() {
  elixir(function (mix) {
    /* Copy Images and font to build folder */
    mix.copy('public/assets/img', 'public/build/img')
      .copy('resources/assets/template/image', 'public/build/image')
      .copy('resources/assets/css/template/stylesheet/fonts', 'public/build/css/fonts')
      .copy('resources/assets/css/fonts', 'public/build/css/fonts');
  });
});

/* 'gulp' - Run for the first time, or when you get the latest */
gulp.task('default',['minifyCSS','minifyJS']);

/* 'gulp watch' - Run only when working on dev */
gulp.task('watch', function () {
  // Watch .js files
  gulp.watch('resources/assets/**/*.js', ['minifyJS']);
  // Watch .css files
  gulp.watch('resources/assets/**/*.css', ['minifyCSS']);
});`
```

Here is my gulp file, I had a fix for version() problem, but I think it's hard to maintain such type of gulp tasks to fix a small issue, Now I'm able to do everything, But I need smaller and simpler version of my gulp file.

**Comment by entreprenik2** [comment]
@a3diti 's [answer](https://github.com/laravel/elixir/issues/126#issuecomment-250315008) fixed it for me.

```
cd public/build/
rm -rf *
```

**Comment by Venu-k** [comment]
I can't delete the files manually for every time, anyhow I fixed the issue in other way. I think the latest elixer has this fix.

**Comment by Sogl** [comment]
Can't run in sequence with external gulp task in latest Elixir:
```js
elixir(function(mix) {

    mix.task('copy')
        .sass(mainSass, 'dist/css/main.css')
        .scripts(vendorScripts, 'dist/js/vendor.js')
        .scripts(mainScripts, 'dist/js/main.js');
});
```
